### ISSUE TITLE
Booting on EC2 fails with "Malformed boot parameter \"ro\""

### ISSUE THREAD

**Issue by sgrove** [mirage/mirage-bootvar-xen#17]
I've made it to the point where I can _almost_ boot a unikernel on EC2, but bootvar doesn't seem to like the boot parameters given to it in this case:

```
Xen Minimal OS!
  start_info: 0xac4000(VA)
    nr_pages: 0x26700
  shared_inf: 0x7decb000(MA)
     pt_base: 0xac7000(VA)
nr_pt_frames: 0x9
    mfn_list: 0x990000(VA)
   mod_start: 0x0(VA)
     mod_len: 0
       flags: 0x0
    cmd_line: root=/dev/sda1 ro 4
  stack:      0x94f860-0x96f860
MM: Init
      _text: 0x0(VA)
     _etext: 0x5ffbd(VA)
   _erodata: 0x78000(VA)
     _edata: 0x80ae0(VA)
stack start: 0x94f860(VA)
       _end: 0x98fe68(VA)
  start_pfn: ad3
    max_pfn: 26700
Mapping memory range 0xc00000 - 0x26700000
setting 0x0-0x78000 readonly
skipped 0x1000
MM: Initialise page allocator for c01000(c01000)-26700000(26700000)
MM: done
Demand map pfns at 26701000-2026701000.
Heap resides at 2026702000-4026702000.
Initialising timer interface
Initialising console ... done.
gnttab_table mapped at 0x26701000.
Initialising scheduler
Thread "Idle": pointer: 0x2026702010, stack: 0x26640000
Initialising xenbus
Thread "xenstore": pointer: 0x20267027c0, stack: 0x26650000
Dummy main: start_info=0x96f960
Thread "main": pointer: 0x2026702f70, stack: 0x26660000
"main" "root=/dev/sda1" "ro" "4" 
vbd 2049 is hd0
******************* BLKFRONT for device/vbd/2049 **********


backend at /local/domain/0/backend/vbd/538/2049
Failed to read /local/domain/0/backend/vbd/538/2049/feature-barrier.
Failed to read /local/domain/0/backend/vbd/538/2049/feature-flush-cache.
2097152 sectors of 512 bytes
**************************
[H[J

    GNU GRUB  version 0.97  (629760K lower / 0K upper memory)



[m[4;2H+-------------------------------------------------------------------------+[5;2H|[5;76H|[6;2H|[6;76H|[7;2H|[7;76H|[8;2H|[8;76H|[9;2H|[9;76H|[10;2H|[10;76H|[11;2H|[11;76H|[12;2H|[12;76H|[13;2H|[13;76H|[14;2H|[14;76H|[15;2H|[15;76H|[16;2H|[16;76H|[17;2H+-------------------------------------------------------------------------+[m

    Use the ^ and v keys to select which entry is highlighted.

    Press enter to boot the selected OS, 'e' to edit the

    commands before booting, or 'c' for a command-line.[5;78H [m[7m[5;3H Mirage                                                                  [5;75H[m[m[6;3H                                                                         [6;75H[m[m[7;3H                                                                         [7;75H[m[m[8;3H                                                                         [8;75H[m[m[9;3H                                                                         [9;75H[m[m[10;3H                                                                         [10;75H[m[m[11;3H                                                                         [11;75H[m[m[12;3H                                                                         [12;75H[m[m[13;3H                                                                         [13;75H[m[m[14;3H                                                                         [14;75H[m[m[15;3H                                                                         [15;75H[m[m[16;3H                                                                         [16;75H[m[16;78H [5;75H[23;4H The highlighted entry will be booted automatically in 1 seconds.   [5;75H[H[J  Booting 'Mirage'



root (hd0)

 Filesystem type is ext2fs, using whole disk

kernel /boot/mirage-os.gz



xc_dom_probe_bzimage_kernel: kernel is not a bzImage
close blk: backend at /local/domain/0/backend/vbd/538/2049
Xen Minimal OS!
Initialising timer interface
Initialising console ... done.
getenv(OCAMLRUNPARAM) -> null
getenv(CAMLRUNPARAM) -> null
getenv(PATH) -> null
Unsupported function lseek called in Mini-OS kernel
Unsupported function lseek called in Mini-OS kernel
Unsupported function lseek called in Mini-OS kernel
getenv(OCAMLRUNPARAM) -> null
getenv(CAMLRUNPARAM) -> null
getenv(TMPDIR) -> null
getenv(TEMP) -> null
getenv(DEBUG) -> null
getenv(OMD_DEBUG) -> null
getenv(OMD_FIX) -> null
Netif: add resume hook

OCaml Xenstore reading vm

OCaml Xenstore reading /vm/ec2952ba-efa5-2e3c-6d22-9b4a59a6fe3c/image/cmdline
Fatal error: exception Failure("Malformed boot parameter \"ro\"")
Raised at file "src/core/lwt.ml", line 789, characters 22-23
Called from file "lib/main.ml", line 58, characters 10-20
Called from file "main.ml", line 274, characters 5-10
Mirage exiting with status 2
Do_exit called!
```

**Comment by jsianes** [comment]
Same issue here using latest commit today. I rolled-back to a previous commit from January (d65242c29031acbf65ff753ca51c316c25ab9330) to make it works for a training session about Unikernels. Any feedback about when the issue will be resolved in the last commit will be much appreciated.

**Comment by amirmc** [comment]
Hi @jsianes, slightly off-topic, but can you give me more info about the training session you ran? (My email is in my GitHub profile).

**Comment by jsianes** [comment]
Hi @amirmc . As part of some internal training sessions, during one of the session I'm introducing Unikernel concept to AWS Support engineers. At the end of the session, we create a 'hello world' AWS AMI using 'mirage-www' source code from Git. The example was working correctly following the steps in your official documentation last November, the first time we ran the session. But testing this week before next session, it seems the Xen unikernel generated with the last commit doesn't work showing "Fatal error: exception Failure("Malformed boot parameter \"ro\"")" at the end of the instance console output. 

As I said in the previous message, rolling-back to a previous commit from January resolve the issue and it's enough as workaround for the next training session. But would be nice if you can fix it for the last stable version or, if there is required any kind of additional configuration, please update the official documentation.

**Comment by talex5** [comment]
For other people with this problem, a work-around is to disable bootvar support with:

`opam pin add -y mirage-bootvar-xen 'https://github.com/talex5/mirage-bootvar-xen.git#qubes'`

(I did this for Qubes, but it sounds like EC2 needs it too)

**Comment by yomimono** [comment]
This should be fixed in version 0.3.2 -- @sgrove , can you confirm when you get a chance?

**Comment by yomimono** [comment]
I'm closing this as it seems resolved to me since #23 and the attendant changes in the `mirage` front-end tool (and I'm now booting unmodified MirageOS unikernels on AWS with the prerelease versions of the tools which have those changes), but please reopen if you find other problems booting on AWS.
